### ISSUE TITLE
Move Process out of BIF/NIF arguments and into a thread-local

### ISSUE THREAD

**Issue by KronicDeth** [lumen/lumen#366]

**Comment by bitwalker** [comment]
# Problem

Currently, we pass around a `Process` reference anywhere that may need to invoke a method on the `Process` type, or call a function that does - most concrete usages are simply taking advantage of the reference to access its heap to allocate on, but such functions don't need to require a `Process` for that. In any case, the result is that the type is infectious, and ends up needing to be passed around to all sorts of code that doesn't care about the concept of a process. 

This makes refactoring a lot more painful, both because it may prevent us from being able to reuse code in places where we don't already pass a process down, or because we need to call such a function deep in a call tree that doesn't have a process involved, triggering a bunch of unnecessary changes, and potentially requiring significant refactors if we are using traits that don't allow us to pass along a Process.

# Solution

The solution here is simple, we define a thread-local global which holds the `Process` active on the current thread, add static versions of the `Process` functions called from BIFs and other areas of the runtime, and dispatch to the `Process` method after loading the reference, e.g.:

```rust
thread_local! {
    static CURRENT_PROCESS: RefCell<Process> = ...;
}

#[inline]
pub fn stack_push(term: Term) -> AllocResult<()> {
    CURRENT_PROCESS.with(|p| p.stack_push(term))
}

impl Process {
    pub fn stack_push(&mut self, term: Term) -> AllocResult<()> { ... }
}
```

As you can see, the static function handles the ceremony of loading the current reference and delegating to the corresponding `Process` method. Callers use `process::stack_push(term)`, rather than `process.stack_push(term)` as they do today.

# Safety

This change is safe due to the following properties:

- All code which invokes a `Process` method is in the same thread as the process it is executing on behalf of. This should always be trivially true, since the only way to be calling such code is on behalf of the `Process` itself. The one thing we need to be aware of is that spawning a new thread to perform some work in on behalf of a process, either cannot call any code which needs to call a `Process` method, OR, must be spawned with its `CURRENT_PROCESS` initialized to the same reference. As far as I'm aware though, we are not doing this anywhere currently, and if such a change is accidentally introduced, it will be trivially caught, as the code would fail as soon as it calls a `Process` method when executing in the new thread.
- Similarly, it is not possible for code to be executing in a thread and have the `CURRENT_PROCESS` change out from underneath it; the code has to yield to the scheduler for such a change to occur, which is precisely how we want it to work

# Effort

The main complication with this change is that the current pattern of passing `Process` everywhere is already ubiquitous, and is only becoming more so, especially with changes such as #363. Making the proposed change will involve:

- Updating every single function signature which has, but doesn't actually need, a `Process` reference to remove that argument. Places that will still use these references primarily only exist in `liblumen_alloc::erts::process`, and in the scheduler implementation.
- Updating every single call to a `Process` method to call the static version instead. This is the most painful change, as it means changing code in virtually every test in `lumen_runtime`, which is an enormous number of files.
- Updating every test that needs a `Process`, to spawn one and make sure the thread-local global is initialized with a reference to it. Luckily this is less invasive, as most of the test setup in `lumen_runtime` is shared across tests, and so this kind of initialization can be done in much fewer places.

-----

In short, this is a painful refactoring, but hugely important for the health of the codebase. Unfortunately, we may need to defer this a short while until codegen is done, at which point we can revisit it - that's assuming those in the community aren't interested in lending a hand here. If that's the case, then I propose that Luke and myself pair up for one or two days when codegen is done and knock this out in one shot.

If anyone is interested in helping with this or taking it on wholesale, and need additional advice or guidance, we're happy to help!

**Comment by KronicDeth** [comment]
This change has been completed since `CURRENT_PROCESS` was implemented in a806e7150e9ae25af26d143c48a0b87ddef2ae2a
